### DISCUSSION TITLE
Publish to Environment - version selection question

### DISCUSSION THREAD

**Discussion by buzzwick** [microsoft/AL-Go#1573] [Q&A]
Data prompt for Publish to Environment says

`App version to deploy to environment(s) (current, prerelease, draft, latest or version number)`

Do these versions apply only to builds of the selected branch  or can they also refer to release names? Could "current" refer both to the most recent version of the app _and_ to a release named "current"?

Additionally, where is the app version number in question located? app.json? 

Can there be more than one "version" of the selected branch?

**Comment by freddydk** [comment]
AL-Go would never expect a release called current - as you would have to delete that and re-create that again and again.
A release should have a version number.

The "magic" words you mention translate to:
- latest is the latest successful build (of the branch in which the Publish is running)
- current is the latest released version (i.e. the current version on the market) - not branch specific
- prerelease is the latest pre-release version - not branch specific
- draft is the latest draft version - not branch specific

The version number is the version number of the artifacts generated (controlled by the repoVersion setting) in the branch in which the publish is running.

![image](https://github.com/microsoft/AL-Go/assets/10775043/3a86a4d9-eecf-434b-9896-bc265724402d)

3.0.432.0 in the above list.

Currently, you cannot deploy an old release to an environment

**Comment by buzzwick** [comment]
Okay, first: "AL-Go would never expect a release called current - as you would have to delete that and re-create that again and again." Hilarious because that is exactly what I have been doing, only using the label "latest" instead of "current." We will stop doing that now...

Second: I did not realize these were "magic" terms--i.e., that they incorporated intelligence about what they are referring to. So there is still some clarification needed. Namely, "latest" refers to a build, whereas all the others refer to releases--is that correct?

Third: So far I only knew about artifacts associated with a release. I see where there is a tab for Artifacts in the Action runs, but I can't seem to find any that actually have Artifacts in them. So there must be only a subset of Actions that actually create build artifacts? Is there a quick way to find what the "latest" artifacts are?

**Comment by freddydk** [comment]
Looks like I should check for these names of releases and throw an error...:-)
On artifacts - only the CI/CD workflow produces useful artifacts.

**Comment by buzzwick** [comment]
So--looking at the parameter prompts for Create Release: if I give a release a name of "v22.2", does it still automatically get the label of "latest" in that green oblong circle? 

And does "current" in the Publish to Environment actually refer to a _Release_ labeled "latest"? Here it seems that "latest" refers to the app builds--NOT--the release with that "latest" label--where all the others refer to Releases. Correct ?

**Comment by freddydk** [comment]
current refers to the latest release - i.e. the release with the highest version number - i.e. the release shown on github:
![image](https://github.com/microsoft/AL-Go/assets/10775043/8001c4ad-91c4-41d8-9814-22d06832d9c9)

**Comment by buzzwick** [comment]
Okay. I get it--I was thrown off by the correspondence between "latest" as a build reference, and "Latest" as an automatically-supplied release label. But now I have your explanations to refer to when I run one of these Actions. Thank you for your patience.

**Comment by freddydk** [comment]
and thanks for pointing out the reason for this misunderstanding
I will also create a chapter in the workshop about releases...

**Comment by buzzwick** [comment]
Yes please!! Obviously I need it! Very welcome!
